[TensorExpr] Do not inline autodiff graphs if they contain prim::TypeCheck nodes.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44564 [TensorExpr] Do not inline autodiff graphs if they contain prim::TypeCheck nodes.**

Before this change we sometimes inlined autodiff subgraph containing
fusion groups. This happened because we didn't look for 'unsupported'
nodes recursively, but fusion groups were inside if-nodes. With this
change we start to look into sub-blocks and also block-list
prim::TypeCheck - if we were to inline those node we would need to
update the expected types and such logic doesn't exist yet.

The problem was detected by @bertmaher in 'LearningToPaint' benchmark
investigation where this bug caused us to keep constantly hitting
fallback paths of the graph.

Differential Revision: [D23657049](https://our.internmc.facebook.com/intern/diff/D23657049)

Commit:
[TensorExpr] Do not inline autodiff graphs if they contain prim::TypeCheck nodes.

Before this change we sometimes inlined autodiff subgraph containing
fusion groups. This happened because we didn't look for 'unsupported'
nodes recursively (maybe we should), but fusion groups were inside
if-nodes.

The problem was detected by @bertrand in 'LearningToPaint' benchmark
investigation where this bug caused us to keep constantly hitting
fallback paths of the graph.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Do not inline autodiff graphs if they contain prim::TypeCheck nodes."


Before this change we sometimes inlined autodiff subgraph containing
fusion groups. This happened because we didn't look for 'unsupported'
nodes recursively (maybe we should), but fusion groups were inside
if-nodes.

The problem was detected by @bertmaher in 'LearningToPaint' benchmark
investigation where this bug caused us to keep constantly hitting
fallback paths of the graph.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/inline_autodiff_subgraphs.cpp:
```diff
@@ -14,7 +14,8 @@ namespace jit {
   auto kind = node->kind();
   return kind != prim::FusionGroup && kind != prim::CudaFusionGroup &&
-      kind != prim::TensorExprGroup && (kind.is_aten() || kind.is_prim());
+      kind != prim::TypeCheck && kind != prim::TensorExprGroup &&
```
Can we change https://github.com/pytorch/pytorch/blob/3fed596cb367e2a8dde858acd5ed5899303de430/torch/csrc/jit/passes/inline_autodiff_subgraphs.cpp#L44 to recurse through the nodes? It's misleading to say that prim::TypeCheck can't be run with Autograd

ZolotukhinM replied:
I added the recursion but also kept `prim::TypeCheck` per your comment below.

eellison reviewed (approved):
I guess if we wanted to be able to inline prim::TypeCheck we would have to update the pass that updates the requires_grad properties of its outputs, so I guess currently it's kind of correct to say that inlining it doesnt preserve the node's behavior

Commit:
Update on "[TensorExpr] Do not inline autodiff graphs if they contain prim::TypeCheck nodes."


Before this change we sometimes inlined autodiff subgraph containing
fusion groups. This happened because we didn't look for 'unsupported'
nodes recursively, but fusion groups were inside if-nodes. With this
change we start to look into sub-blocks and also block-list
prim::TypeCheck - if we were to inline those node we would need to
update the expected types and such logic doesn't exist yet.

The problem was detected by @bertmaher in 'LearningToPaint' benchmark
investigation where this bug caused us to keep constantly hitting
fallback paths of the graph.

Differential Revision: [D23657049](https://our.internmc.facebook.com/intern/diff/D23657049)

[ghstack-poisoned]
